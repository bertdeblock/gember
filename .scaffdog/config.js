export default {
  files: ["documents/*.md"],
  tags: ["[[", "]]"],
};
