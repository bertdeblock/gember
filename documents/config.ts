import type { Config } from "scaffdog";

/**
 * Given an array of strings, this function will return a tuple of two arrays:
 * - The first array will contain all the strings that do not begin with a '!'.
 * - The second array will contain all the strings that do begin with a '!',
 *   with the '!' removed.
 */
function parseOptions(options: string[]): [string[], string[]] {
  const positive: string[] = [];
  const negative: string[] = [];

  for (let i = 0; i < options.length; i++) {
    if (options[i].startsWith("!")) {
      negative.push(options[i].slice(1));
    } else {
      positive.push(options[i]);
    }
  }

  return [positive, negative];
}

const config: Config = {
  files: ["*.md"],
  helpers: [
    (registry): void => {
      /**
       * This helper looks at `context.inputs` and `options` and returns either
       * '' or '!' based on the values of the options in `context.inputs`.  If
       * an option starts with a '!', it is considered a negative option and
       * will be checked for falsy values. Otherwise, the option will be
       * checked for truthy values.
       *
       * For example, if `context.inputs` is `{ foo: true, bar: false }`, then:
       * - `{{hasOpts "foo"}}` will return '' because `foo` is truthy
       * - `{{hasOpts "!foo"}}` will return '!' because `foo` is truthy
       * - `{{hasOpts "bar"}}` will return '!' because `bar` is falsy
       * - `{{hasOpts "!bar"}}` will return '' because `bar` is falsy
       * - `{{hasOpts "foo" "!bar"}}` will return '' because `foo` is truthy
       *      and `bar` is falsy
       * - `{{hasOpts "foo" "bar"}}` will return '!' because `foo` is truthy
       *      and `bar` is falsy
       * - `{{hasOpts "!foo" "!bar"}}` will return '!' because `foo` is truthy
       *      and `bar` is falsy
       */
      registry.set("hasOpts", (context, ...options) => {
        const [positive, negative] = parseOptions(options);
        const inputs = context.variables.get("inputs");

        const allPositive = positive.every((opt) => inputs?.[opt]);
        const noNegative = negative.every((opt) => !inputs?.[opt]);

        return allPositive && noNegative ? "" : "!";
      });
    },
  ],
};
export default config;
