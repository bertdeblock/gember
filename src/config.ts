import { deepmerge } from "deepmerge-ts";
import { findUp } from "find-up";
import { pathToFileURL } from "node:url";

type Config = {
  generator: {
    component: { path: string };
    helper: { path: string };
    modifier: { path: string };
    service: { path: string };
  };
  ts: boolean;
};

const DEFAULT_CONFIG: Config = {
  generator: {
    component: { path: "" },
    helper: { path: "" },
    modifier: { path: "" },
    service: { path: "" },
  },
  ts: false,
};

const CONFIG_FILES = [
  "gember.config.js",
  "gember.config.cjs",
  "gember.config.mjs",
];

export async function getConfig(cwd: string): Promise<Config> {
  const path = await findUp(CONFIG_FILES, { cwd });

  if (path === undefined) {
    return DEFAULT_CONFIG;
  }

  let config;

  try {
    config = (await import(pathToFileURL(path).toString())).default;
  } catch {
    throw new Error(`Could not import gember config file at "${path}".`);
  }

  if (config === undefined) {
    throw new Error(
      `gember config file at "${path}" must have a "default" export.`,
    );
  }

  return deepmerge(
    DEFAULT_CONFIG,
    typeof config === "function" ? await config() : config,
  );
}
