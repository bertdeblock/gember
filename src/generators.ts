import { generate } from "./generate.js";

export function generateComponent(
  name: string,
  packagePath: string,
  {
    classBased = false,
    nested,
    path,
    typescript = false,
  }: {
    classBased?: boolean;
    nested?: boolean;
    path?: string;
    typescript?: boolean;
  } = {},
): Promise<void> {
  return generate({
    customTargetPath: path,
    entityName: name,
    generatorName: "component",
    nested,
    packagePath,
    targetDir: "components",
    templateFilename:
      (classBased ? "class-based" : "template-only") +
      (typescript ? ".gts" : ".gjs"),
  });
}

export function generateHelper(
  name: string,
  packagePath: string,
  {
    classBased = false,
    path,
    typescript = false,
  }: {
    classBased?: boolean;
    path?: string;
    typescript?: boolean;
  } = {},
): Promise<void> {
  return generate({
    customTargetPath: path,
    entityName: name,
    generatorName: "helper",
    packagePath,
    targetDir: "helpers",
    templateFilename:
      (classBased ? "class-based" : "function-based") +
      (typescript ? ".ts" : ".js"),
  });
}

export function generateModifier(
  name: string,
  packagePath: string,
  {
    classBased = false,
    path,
    typescript = false,
  }: {
    classBased?: boolean;
    path?: string;
    typescript?: boolean;
  } = {},
): Promise<void> {
  return generate({
    customTargetPath: path,
    entityName: name,
    generatorName: "modifier",
    packagePath,
    targetDir: "modifiers",
    templateFilename:
      (classBased ? "class-based" : "function-based") +
      (typescript ? ".ts" : ".js"),
  });
}

export function generateService(
  name: string,
  packagePath: string,
  {
    path,
    typescript = false,
  }: {
    path?: string;
    typescript?: boolean;
  } = {},
): Promise<void> {
  return generate({
    customTargetPath: path,
    entityName: name,
    generatorName: "service",
    packagePath,
    targetDir: "services",
    templateFilename: typescript ? "service.ts" : "service.js",
  });
}
