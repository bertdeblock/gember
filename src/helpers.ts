import type { EmberPackageJson } from "./types.js";

export function isV1Addon(packageJson: EmberPackageJson): boolean {
  if (isAddon(packageJson)) {
    const { version } = packageJson["ember-addon"] ?? {};

    return version === 1 || version === undefined;
  }

  return false;
}

export function isV2Addon(packageJson: EmberPackageJson): boolean {
  if (isAddon(packageJson)) {
    const { version } = packageJson["ember-addon"] ?? {};

    return version === 2;
  }

  return false;
}

function isAddon(packageJson: EmberPackageJson): boolean {
  if (Array.isArray(packageJson.keywords)) {
    return packageJson.keywords.includes("ember-addon");
  }

  return false;
}
