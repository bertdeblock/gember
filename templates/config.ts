import type { Config } from "scaffdog";

const config: Config = {
  files: ["*.md"],
};

export default config;
