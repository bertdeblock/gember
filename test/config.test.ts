import { it } from "vitest";
import { resolveConfig } from "../src/config.js";
import { generateComponent } from "../src/generators.ts";
import { gember, Package } from "./helpers.js";

it("supports a `gember.config.js` file", async (ctx) => {
  const pkg = await Package.create("v2-addon-config-js");

  const config = await resolveConfig(pkg.path);

  ctx.expect(config).to.deep.equal({ hooks: {} });
});

it("supports a `gember.config.cjs` file", async (ctx) => {
  const pkg = await Package.create("v2-addon-config-cjs");

  const config = await resolveConfig(pkg.path);

  ctx.expect(config).to.deep.equal({ hooks: {} });
});

it("supports a `gember.config.mjs` file", async (ctx) => {
  const pkg = await Package.create("v2-addon-config-mjs");

  const config = await resolveConfig(pkg.path);

  ctx.expect(config).to.deep.equal({ hooks: {} });
});

it("runs the `postGenerate` hook", async (ctx) => {
  const pkg = await Package.create("v2-addon-config", "post-generate-info");

  await generateComponent("foo", pkg.path);

  const content = await pkg.readFile("post-generate-info.json");

  ctx.expect(content).toMatchSnapshot();

  await pkg.cleanUp();
});

it("applies specific generator options", async (ctx) => {
  const pkg = await Package.create("v2-addon-config");

  await gember(["component", "foo"], { cwd: pkg.path });

  const content = await pkg.readFile("src/components/foo.gts");

  ctx.expect(content).toMatchSnapshot();

  await pkg.cleanUp();
});
