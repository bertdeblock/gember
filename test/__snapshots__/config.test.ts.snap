// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runs the \`postGenerate\` hook 1`] = `
"{
  "documentName": "component",
  "entityName": "foo",
  "files": [
    {
      "content": "<template>{{yield}}</template>\\n",
      "name": "foo.gjs",
      "path": "post-generate-info/src/components/foo.gjs"
    }
  ]
}"
`;
